custom chalkboard key config added (main.js + rise.yaml)

theck3r opened the pull request:
Hi,
I have added some code to customize the keyboard short cuts for the chalkboard (as most of the hard wired default keys are not available on my german keyboard).

I guess this is not the perfect solution (read the discussion in #442), but it allows using some custom hotkey fields in nbconfig and converts the key-strings into key codes, which are then passed to the RevealChalkboard setup (could be used for other reveal.js keys as well, I guess...).

Conversion is done by a pretty simple string concatenation of the nbconfig key-shortcut string(s) to KeyEvent's VirtualKeyboardConstants (DOM_VK_...) - falls back to the default key codes, if no match is found.
I also added some code for updating the Rise help accordingly - so the help menu should always show the user defined shortcuts (if they could be converted).

Seems to work well on my system - linux/firefox (although some DOM_VK_... constants are missing), but I am rather new to javascript, so any feedback is welcome.

regards,
Thies

parmentelat commented:
Hi

That sounds pretty exciting, and I've wished I had a way to customize hard-wired shortcuts for a long time now; and I can see you've spotted the original discussion on that matter

I had other ideas about how to deal with that, but that's going to be for some other time - if ever, actually..

so on the principle I'm fine; I'm not sure when I will have a chance to release this; our plans were to focus on the jlab thing, but that has gotten a little stuck in the meantime, so I guess the classic notebook extension still has some future; I guess it would belong in a 5.6.1.dev<n>

if it's not too much hassle, I'd love if you could extend this patch to cover the `home` and `end` functions as well among the parametrizable features ?
also, do you see any impact on the documentation as shown at rise.readthedocs.io ? 

thanks !

theck3r commented:
I have just pushed some updates including the extension for the `home` and `end` keys you asked for. Actually I tried to cover all bindings that were directly assigning keycodes to reveal.js API calls or RISE functions (so I have included e.g. toggleOverview, toggleAllRiseButtons,...).

To add some more structure I have now split up the RISE/reveal.js and the chalkboard bindings in nbconfig. I have added a short chapter to the documentation (good hint), which I hope explains the idea.
Additionally I have moved some code blocks into functions to make the whole thing a bit more versatile. So adding custom key bindings to further API calls should now be easier.

If we would only support to make the native key customization via the nbconfig JSON (not `nbextension_configurator`) the whole thing could easily be simplified and would be more flexible (i.e. just define a list of dicts with keyboard shortcut and API call). But of course this is not as intuitive to use as having predefined bindings in the configurator and if there is not many requests for further customized reveal.js bindings, it's probably not that interesting.

parmentelat commented on rise/static/main.js:
```diff
@@ -526,13 +627,18 @@ define([
                   40: null, // down arrow disabled
                   66: null, // b, black pause disabled, use period or forward slash
-                  70: fullscreenHelp, // disable fullscreen inside the slideshow, makes codemirror unreliable
+                  // 70: fullscreenHelp, // disable fullscreen inside the slideshow, makes codemirror unreliable
```
IIRC the initial purpose of use redefining that 70 key - I guess it's 'f' or something - is to disable the native 'fullscreen' reveal feature 
(to be honest it was disabled a long time ago with an older version of reveal; and I'm not sure if it's still a problem or not, but that's another thing entirely of course)
so IMHO it would be safer to keep `70: null` in this list, to be sure to disable that binding, even if the user decides to map that function to another key

theck3r replied:
I was also a bit confused by this action... But again I agree - this list should contain only disabling of bindings and default/custom bindings will be assigned in the next step.

parmentelat commented:
as I said I'm going to need to find some time to play with this, but it sounds promising indeed 
thanks !

theck3r commented on doc/customize.md:
```diff
@@ -499,3 +499,57 @@ Note that with this approach, you will end up with the
+         }
+    }
+
+Note, that it is not possible to define key combinations (e.g. `Alt-C`) for 
```
As far as I understand the [reveal.js documentation regarding keyboard bindings](https://github.com/hakimel/reveal.js/#keyboard-bindings) they only attach to a keycode (corresponding to a keydown event, if I am correct). So this limits bindings to single key presses from reveal.js side.
The RISE source code already contained a comment to analyze, if API calls to RevealChalkboard could be managed directly by jupyter w.o. using the binding interface. Seems like a good idea, but requires some more knowledge on jupyter and I didn't dig any deeper into that (yet)...

theck3r commented on rise/static/main.js:
```diff
@@ -541,8 +647,8 @@ define([
-                  87: Reveal.toggleOverview, // w, toggle overview
-                  188: toggleAllRiseButtons, // comma
+                  // 87: Reveal.toggleOverview, // w, toggle overview
+                  // 188: toggleAllRiseButtons, // comma
```
good point - I agree.

parmentelat commented:
it's a little embarrassing because our focus is not the `master` branch but the `maint-5.6` instead  
my bad entirely, this is not documented anywhere..

anyways, I've merged, I'll see how it goes

parmentelat commented:
all right, so here's where i am
* I merged your contribution into master
* I merged this into maint-5.6 which as I said is our current focus; no worries, it was just a little off
* I took the liberty to reindent the whole js file, mostly because you were not following our conventions; 
* also I moved around your additions in `rise.yaml` to come after the original section on keystrokes

----
at this point what I see as remaining issues is 

* it's my feeling that the yaml file should come with more defaults; I mean for exemple we have 'firstSlide' bound to 'home'  - or at least we used to have, and it should remain so
so we should IMHO have that reflected in the yaml file as a default for `rise.reveal_shortcuts.main.firstSlide`

* I have only made a quick test, I can't see anything broken but I was not able to configure things through the configurator - which honestly is not conclusive as I have recently changed a lot of my environment, so I can't tell for sure where the configurator is saving things and whether it affects my dev. miniconda or not...

in any case, if you plan to make any further change, please start from the maint-5.6 branch;
feel free to open another PR in that case, as I think it's the most convenient way given that change

thanks !

parmentelat commented:
@damianavila 
it'd help if you were able to do some tests on this on your end

damianavila commented:
@theck3r thank you very much for this work!
